sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -509,16 +509,25 @@ case class In(value: Expression, list: Seq[Expression]) extends Predicate {
     val valueArg = ctx.freshName("valueArg")
     // All the blocks are meant to be inside a do { ... } while (false); loop.
     // The evaluation of variables can be stopped when we find a matching value.
-    val listCode = listGen.map(x =>
-      s"""
-         |${x.code}
-         |if (${x.isNull}) {
-         |  $tmpResult = $HAS_NULL; // ${ev.isNull} = true;
-         |} else if (${ctx.genEqual(value.dataType, valueArg, x.value)}) {
-         |  $tmpResult = $MATCHED; // ${ev.isNull} = false; ${ev.value} = true;
-         |  continue;
-         |}
-       """.stripMargin)
+    val listCode = listGen.map { x =>
+      val codeIf =
+        s"""
+           |${x.code}
+           |if (${x.isNull}) {
+           |  $tmpResult = $HAS_NULL; // ${ev.isNull} = true;
+           |}
+         """.stripMargin
+      val codeElseIf =
+        if (!java.lang.Boolean.parseBoolean(x.isNull.toString)) {
```

Comment:
better to add a comment here

Reply:
`x.null` is a `ExprCode` type. and toString returns code string. So, there are three cases:
a. code is a variable name of the Boolean type, for example, if (varArg_0)
b. code is false, if (false)
c. code is true, if (true)
The `else if` branch does not need to be generated when isNull is set to true.
 
![image](https://github.com/apache/spark/assets/132866841/9e133749-30f7-4644-8d56-ede17bea7ff4)